addons/sale_stock/models/sale_order.py

```diff
@@ -364,14 +350,25 @@ def create(self, vals_list):
 
     def write(self, values):
         lines = self.env['sale.order.line']
+        lines_after_return = self.env['sale.order.line']
         if 'product_uom_qty' in values:
             precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
-            lines = self.filtered(
-                lambda r: r.state == 'sale' and not r.is_expense and float_compare(r.product_uom_qty, values['product_uom_qty'], precision_digits=precision) == -1)
+            for line in self:
+                if line.state != 'sale' or line.is_expense:
+                    continue
+                if float_compare(line.product_uom_qty, values['product_uom_qty'], precision_digits=precision) == -1:
+                    lines |= line
+                elif line.qty_delivered > 0:
+                    lines_after_return |= line
```

Comment:
You should probably check if there is a backorder and in this case reduce the quantity on it or create a new procurement with
order_line.product_uom_qty - order_line.qty_delivered - existing inital demand on open BO